src/aead/aead.rs

```diff
@@ -77,11 +77,14 @@ impl OpeningKey {
 
 /// Authenticates and decrypts (“opens”) data in place. When
 ///
-/// The input may have a prefix that is `in_prefix_len` bytes long; any such
+/// `nonce` must be unique for every use of the key to seal data; it must be
```

Comment:
> I was going to suggest that when you wrote "I don't recommend using random nonces with these AEADs" over here #421 (comment), which was a total surprise to me.

See also https://boringssl-review.googlesource.com/c/13440/#message-e6184e6b129cf0ac487dbe18c0ae85c54174bed1 and https://www.blackhat.com/docs/us-16/materials/us-16-Devlin-Nonce-Disrespecting-Adversaries-Practical-Forgery-Attacks-On-GCM-In-TLS-wp.pdf.

> As it's an actual security concern and I wouldn't expect anyone except absolute crypto experts to know about this, it's something that the docs should tell you.

Maybe. Right now *ring*'s documentation isn't intended to teach people cryptography, just intended to teach people how to use *ring*.

> I get your point. It's from the viewpoint of someone who uses ring to implement a given protocol.

This is what *ring* is designed for.

> As you may have noticed, my viewpoint is usually someone who just tries to encrypt something securely and send it accross the network. If you will, such users are designing their own "protocol" (albeit a tiny one) and for their protocol design, they have to understand the fundamental concept that the nonce needs to be the same on both ends.

If they don't understand already that the nonce would need to be the same at both ends, they shouldn't be designing their own protocol.

> but for all the user knows about opening, they could again specify a new, unique and random nonce.

That is true, but then their thing won't work. And if they choose the wrong nonce then there isn't any security issue because AES-GCM is designed to fail safe when the wrong nonce is used during opening, and also because *ring* fails safe when things go wrong (e.g. overwriting the ciphertext/plaintext when authentication fails, which is what would happen with a wrong nonce or wrong key or wrong ciphertext or wrong AAD).

> That's especially the case when these users come from using AES-ECB encryption, and someone told them it's insecure and they have to use AEAD instead. They don't have to change that much, but they have to know about tags and nonces now.

They literally need to read a book. *ring*'s documentation isn't intended to replace a book or other educational tools about designing protocols. Rather, *ring* is designed to make it easy to verify that a *ring*-based implementation matches a specification of the higher-level protocol being implemented.